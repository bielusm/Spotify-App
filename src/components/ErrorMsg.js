import React from "react";
import { connect } from "react-redux";

export const ErrorMsg = props => (
  <>
    {props.errMsg && (
      <h2 id="errMsg" className="my-4">
        {props.errMsg}
      </h2>
    )}
  </>
);

const mapStateToProps = state => ({ errMsg: state.error });

export default connect(mapStateToProps)(ErrorMsg);
