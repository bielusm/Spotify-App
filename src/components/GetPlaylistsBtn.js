import React from "react";

const GetPlaylistsBtn = props => (
  <button
    id="getPlaylistsBtn"
    className="button"
    onClick={props.getPlaylists}
    disabled={props.getPlaylistsDisabled}
  >
    Pick playlists
  </button>
);

export default GetPlaylistsBtn;
