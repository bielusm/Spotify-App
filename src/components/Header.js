import React from "react";

const Header = () => <h1>Spotify Manager</h1>;
export default Header;
