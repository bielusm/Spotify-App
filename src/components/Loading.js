import React from "react";

export const Loading = props => (
  <div>
    {props.loading && (
      <div className="loadingStatus">
        <div className="spinner-border" role="status">
          <span className="sr-only"></span>
        </div>
        <h2 className="mx-2 d-inline">Loading...</h2>
      </div>
    )}
  </div>
);

export default Loading;
