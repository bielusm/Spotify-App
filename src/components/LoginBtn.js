import React from "react";

export default class LoginBtn extends React.Component {
  login = () => {
    const scope =
      "playlist-modify-public%20user-read-playback-state%20playlist-read-private%20playlist-modify-private";
    const redirect_uri = window.location.origin;
    //redirect_uri = redirect_uri.slice(0, redirect_uri.length - 1);
    const url = `https://accounts.spotify.com/authorize?client_id=4252feb807d04ced962e15f346258957&response_type=token&redirect_uri=${redirect_uri}&scope=${scope}&show_dialog=false`;
    window.location.href = url;
  };

  render() {
    return (
      <div>
        {this.props.loginVisible && (
          <button id="login" className="btn btn-secondary" onClick={this.login}>
            Log in
          </button>
        )}
      </div>
    );
  }
}
