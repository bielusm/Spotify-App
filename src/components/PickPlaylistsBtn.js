import React from "react";
import Button from "@material-ui/core/Button";

export const PickPlaylistsBtn = props => (
  <Button
    variant="contained"
    color="primary"
    id="pickPlaylistsBtn"
    onClick={props.getPlaylists}
    disabled={props.getPlaylistsDisabled}
  >
    Pick playlists
  </Button>
);

export default PickPlaylistsBtn;
