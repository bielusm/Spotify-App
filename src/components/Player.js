import React from "react";
import { Button, IconButton } from "@material-ui/core";

import {
  modifyLibrary,
  startPlayer,
  pausePlayer,
  nextTrack,
  prevTrack
} from "../api/player";

import { setErrorMsg } from "../store/actions/error";
import { connect } from "react-redux";

export class Player extends React.Component {
  togglePause = async () => {
    if (this.props.is_playing)
      pausePlayer(this.props.bearerToken).then(() => this.props.update());
    else startPlayer(this.props.bearerToken).then(() => this.props.update());
  };

  nextTrack = async () => {
    nextTrack(this.props.bearerToken).then(() => this.props.update());
  };

  prevTrack = async () => {
    prevTrack(this.props.bearerToken).then(() => {
      this.props.update();
    });
  };

  toggleLike = async () => {
    modifyLibrary(
      this.props.bearerToken,
      this.props.id,
      this.props.likes_song
    ).then(() => this.props.update());
  };
  render() {
    return (
      <>
        <IconButton
          className="prevBtn fas fa-backward"
          onClick={this.prevTrack}
        />

        <IconButton
          className={
            "playBtn fas " + (this.props.is_playing ? "fa-pause" : "fa-play")
          }
          onClick={this.togglePause}
        />
        <IconButton
          className="nextBtn fas fa-forward"
          onClick={this.nextTrack}
        />
        <IconButton
          className={
            "likeBtn " +
            (this.props.likes_song ? "fas fa-heart" : "far fa-heart")
          }
          onClick={this.toggleLike}
        />
      </>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setErrorMsg: error => dispatch(setErrorMsg(error))
});

export default connect(undefined, mapDispatchToProps)(Player);
