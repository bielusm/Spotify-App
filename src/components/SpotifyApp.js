import React from "react";

import Header from "./Header";
import LoginBtn from "./LoginBtn";
import UpdateBtn from "./UpdateBtn";
import GetPlaylistsBtn from "./GetPlaylistsBtn";
import ErrorMsg from "./ErrorMsg";
import TrackContext from "./TrackContext";
import Loading from "./Loading";
import PlaylistSelection from "./PlaylistSelection";
import TrackedPlaylists from "./TrackedPlaylists";
import Player from "./Player";

import { currentPlayer, likesSong } from "../api/player";
import Playlists from "../api/playlists";

export default class SpotifyApp extends React.Component {
  state = {
    loginVisible: true,
    trackContext: "",
    showPlaylistsDiv: true,
    showTrackedPlaylists: false,
    showPlaylists: false,
    errMsg: undefined,
    playlists: undefined,
    trackedPlaylists: [],
    loading: false
  };
  bearerToken = null;
  currentTrack = null;

  updateTimer = null;
  playlists = null;
  is_playing = false;
  likesSong = false;
  /**
   * Updates the UI
   * @async
   */
  update = async () => {
    try {
      //need to better parall this
      const newTrack = await currentPlayer(this.bearerToken);
      this.likesSong = await likesSong(this.bearerToken, newTrack.id);
      this.is_playing = newTrack.is_playing;
      if (
        this.currentTrack === null ||
        this.currentTrack.uri !== newTrack.Uri
      ) {
        this.setTrackContext(newTrack);
      }
      this.markInPlaylist();
      if (this.updateTimer === null)
        this.updateTimer = setInterval(this.update, 2000);
      this.success();
    } catch (msg) {
      this.error(msg);
    }
  };

  /**
   * Takes a spotify track object and displays the name and artist to the user
   *
   * @param {Spotify Track Object} currentTrack
   */
  setTrackContext = currentTrack => {
    this.currentTrack = currentTrack;
    let trackContext = currentTrack.name + " by ";
    currentTrack.artists.forEach(artist => {
      trackContext += " " + artist.name + ",";
    });
    //remove trailing comma
    trackContext = trackContext.slice(0, trackContext.length - 1);
    document.title = trackContext;
    this.setState(() => ({ trackContext }));
  };

  /**
   * gets a list of all users playlists and prints them on the DOM
   *
   * @async
   */
  getPlaylists = async () => {
    let showPlaylists = !this.state.showPlaylists;

    if (this.state.playlists === undefined) {
      try {
        const playlistsFetch = await this.playlists.getPlaylists(
          this.bearerToken
        );

        const playlists = playlistsFetch.items.map(item => ({
          ...item,
          selected: false
        }));
        this.setState({
          playlists
        });
        this.success();
      } catch (msg) {
        this.error(msg);
        //do not show playlists on error
        showPlaylists = false;
      }
    }

    this.setState(() => ({ showPlaylists }));
    this.addNewPlaylists();
  };

  /**
   * Adds new tracked playlists to the playlist array in the API class
   */
  addNewPlaylists = () => {
    if (this.state.trackedPlaylists.length > 0) {
      let promises = [];

      this.setState(() => ({
        getPlaylistsDisabled: true,
        updateDisabled: true,
        showTrackedPlaylists: true,
        loading: true
      }));

      this.state.trackedPlaylists.forEach(playlist => {
        if (playlist.loading) {
          promises.push(
            this.playlists.addPlaylistByID(
              this.bearerToken,
              playlist.id,
              playlist.name
            )
          );
        }
      });
      Promise.all(promises)
        .then(() => {})
        .catch(msg => error(msg))
        .finally(() => {
          this.setState(prevState => ({
            loading: false,
            trackedPlaylists: prevState.trackedPlaylists.map(pl => {
              pl.loading = false;
              return pl;
            })
          }));
        });
    }
  };

  //When the user clicks on a playlist it will become selected/unselected
  selectPlaylist = e => {
    try {
      e.persist();
      const { tagName, id, classList } = e.target;
      if (classList.contains("playlist-selection__item")) {
        const pl = this.state.playlists.find(pl => pl.id === id);
        const selected = !pl.selected;
        if (selected === true)
          this.setState(prevState => ({
            trackedPlaylists: [
              ...prevState.trackedPlaylists,
              {
                name: e.target.textContent,
                id: e.target.id,
                found: false,
                loading: true
              }
            ]
          }));
        else {
          this.setState(prevState => {
            return {
              trackedPlaylists: prevState.trackedPlaylists.filter(
                item => item.id !== e.target.id
              )
            };
          });
          this.playlists.removePlaylist(e.target.id);
        }

        this.setState(prevState => ({
          playlists: prevState.playlists.map(pl => {
            if (pl.id === id) return { ...pl, selected: !pl.selected };
            else return pl;
          })
        }));
      }
    } catch (err) {
      this.error(err);
    }
  };

  /**
   * Asks the API if current track is in playlist, if it is display that to the user,
   * adds remove and add buttons to each playlist in the tracked playlist list
   */
  markInPlaylist = () => {
    const inPlaylists = this.playlists.trackInPlaylists(this.currentTrack);
    const trackedPlaylists = this.state.trackedPlaylists;

    trackedPlaylists.forEach(child => {
      child.found = false;
    });

    inPlaylists.forEach(inPl => {
      const pl = trackedPlaylists.find(element => element.name === inPl.name);
      pl.found = true;
    });
  };

  addOrRemove = async e => {
    if (e.target.tagName === "BUTTON") {
      e.persist();
      e.target.disabled = true;
      const playlist_id = e.target.parentElement.id;
      let track_uri = this.currentTrack.uri;
      let promise = null;
      try {
        if (e.target.classList.contains("remove")) {
          await this.playlists.removeTrackFromPlaylist(
            this.bearerToken,
            playlist_id,
            this.currentTrack
          );
        } else {
          await this.playlists.addTrackToPlaylist(
            this.bearerToken,
            playlist_id,
            this.currentTrack
          );
        }
        this.update().finally(() => {
          e.target.disabled = false;
        });
      } catch (err) {
        this.error(err);
      }
    }
  };

  //clears the error msg on success
  success = () => {
    this.setState(() => ({ errMsg: undefined }));
  };

  //handles error msgs given from rejected promises
  error = error => {
    console.log(error);
    const errLoc = document.querySelector("#errMsg");
    let msg = error.message;
    if (!isNaN(msg)) {
      const status = parseInt(msg);
      switch (status) {
        case 401:
          msg = "Please login, sessions are only valid for one hour";
          this.setState(() => ({ loginVisible: true }));
          break;
        case 429:
          msg =
            "Error: too many requests to spotify api, please wait a little and try again";
          break;
        default: {
          msg = status;
        }
      }
    }
    this.setState(() => ({ errMsg: msg }));
  };

  componentDidMount() {
    let hash = location.hash;

    if (location.hash !== "") {
      this.setState(() => ({ loginVisible: false }));
    }
    hash = hash.replace("#", "?");
    const urlParams = new URLSearchParams(hash);

    this.bearerToken = urlParams.get("access_token");
    this.playlists = new Playlists();
  }

  render() {
    return (
      <div>
        <div className="container">
          <Header />
          <div className="button-container">
            <LoginBtn loginVisible={this.state.loginVisible} />
            <UpdateBtn update={this.update} />
            <GetPlaylistsBtn
              getPlaylists={this.getPlaylists}
              disabled={this.state.getPlaylistsDisabled}
            />
          </div>
          <ErrorMsg msg={this.state.errMsg} />
          <TrackContext trackContext={this.state.trackContext} />

          {this.currentTrack !== null && (
            <Player
              bearerToken={this.bearerToken}
              is_playing={this.is_playing}
              likes_song={this.likesSong}
              id={this.currentTrack.id}
              update={this.update}
            />
          )}
          <Loading loading={this.state.loading} />
          <PlaylistSelection
            showPlaylists={this.state.showPlaylists}
            playlists={this.state.playlists}
            selectPlaylist={this.selectPlaylist}
          />
          <TrackedPlaylists
            trackedPlaylists={this.state.trackedPlaylists}
            currentTrack={this.currentTrack}
            addOrRemove={this.addOrRemove}
          />
        </div>
      </div>
    );
  }
}
