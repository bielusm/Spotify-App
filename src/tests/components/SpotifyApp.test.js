import React from "react";
import { shallow } from "enzyme";
import { SpotifyApp } from "../../components/SpotifyApp";
import { JestEnvironment } from "@jest/environment";

test("SpotifyApp snapshot test", () => {
  const wrapper = shallow(<SpotifyApp setAccessToken={jest.fn()} />);
  expect(wrapper).toMatchSnapshot();
});
