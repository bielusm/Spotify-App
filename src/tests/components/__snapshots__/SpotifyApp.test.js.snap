// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotifyApp snapshot test 1`] = `
<div>
  <div
    className="container"
  >
    <Header />
    <div
      className="button-container"
    >
      <LoginBtn
        loginVisible={true}
      />
      <UpdateBtn
        update={[Function]}
      />
      <GetPlaylistsBtn
        getPlaylists={[Function]}
      />
    </div>
    <Connect(ErrorMsg) />
    <TrackContext
      trackContext=""
    />
    <Loading
      loading={false}
    />
    <PlaylistSelection
      selectPlaylist={[Function]}
      showPlaylists={false}
    />
    <TrackedPlaylists
      addOrRemove={[Function]}
      currentTrack={null}
      trackedPlaylists={Array []}
    />
  </div>
</div>
`;
